src/js_native_api_v8.cc

```diff
@@ -485,105 +494,62 @@ RefBase* RefBase::New(napi_env env,
 }
 
 uint32_t RefBase::Ref() {
-  return ++_refcount;
+  return ++refcount_;
 }
 
 uint32_t RefBase::Unref() {
-  if (_refcount == 0) {
+  if (refcount_ == 0) {
     return 0;
   }
-  return --_refcount;
+  return --refcount_;
 }
 
 uint32_t RefBase::RefCount() {
-  return _refcount;
-}
-
-void RefBase::Finalize(bool is_env_teardown) {
-  // In addition to being called during environment teardown, this method is
-  // also the entry point for the garbage collector. During environment
-  // teardown we have to remove the garbage collector's reference to this
-  // method so that, if, as part of the user's callback, JS gets executed,
-  // resulting in a garbage collection pass, this method is not re-entered as
-  // part of that pass, because that'll cause a double free (as seen in
-  // https://github.com/nodejs/node/issues/37236).
-  //
-  // Since this class does not have access to the V8 persistent reference,
-  // this method is overridden in the `Reference` class below. Therein the
-  // weak callback is removed, ensuring that the garbage collector does not
-  // re-enter this method, and the method chains up to continue the process of
-  // environment-teardown-induced finalization.
-
-  // During environment teardown we have to convert a strong reference to
-  // a weak reference to force the deferring behavior if the user's finalizer
-  // happens to delete this reference so that the code in this function that
-  // follows the call to the user's finalizer may safely access variables from
-  // this instance.
-  if (is_env_teardown && RefCount() > 0) _refcount = 0;
-
-  if (_finalize_callback != nullptr) {
-    // This ensures that we never call the finalizer twice.
-    napi_finalize fini = _finalize_callback;
-    _finalize_callback = nullptr;
-    _env->CallFinalizer(fini, _finalize_data, _finalize_hint);
-  }
-
-  // this is safe because if a request to delete the reference
-  // is made in the finalize_callback it will defer deletion
-  // to this block and set _delete_self to true
-  if (_delete_self || is_env_teardown) {
-    Delete(this);
-  } else {
-    _finalize_ran = true;
+  return refcount_;
+}
+
+void RefBase::Finalize() {
+  napi_finalize finalize_callback = finalize_callback_;
+  bool delete_self = delete_self_;
+
+  // Either the RefBase is going to be deleted in the finalize_callback or not,
+  // it should be removed from the tracked list.
+  Unlink();
+  // 1. If the finalize_callback is present, it should either delete the
+  //    RefBase, or set the flag `delete_self`.
+  // 2. If the finalizer is not present, the RefBase can be deleted after the
+  //    call.
+  if (finalize_callback != nullptr) {
+    env_->CallFinalizer(finalize_callback, finalize_data_, finalize_hint_);
+    // No access to `this` after finalize_callback is called.
+  }
+
+  // If the RefBase is not self-destructive, userland code should delete it.
+  // Now delete it if it is self-destructive.
+  if (delete_self) {
+    delete this;
```

Comment:
Is it possible to get to a situation when we delete the object at the same address twice? What if some other object allocated at the same address after the first delete? 
It feels that it would be safer to avoid Reference deletion while we are in its Finalizer callback.
E.g. add a flag that we are in a finalizer, and if so, then just record the deletion intent by setting the `delete_self_`  field. Then, after exiting the finalizer we know that the Reference is still alive and act on its `delete_self_` field instead of the locally stored `delete_self` variable.

Reply:
`delete_self_` flag is used to indicate that the data is owned by node-api and node-api should delete the `RefBase` once the data has been finalized.